File: app/src/main/java/com/pr0gramm/app/ui/dialogs/UpdateDialogFragment.java
package com.pr0gramm.app.ui.dialogs;

import android.app.Dialog;
import android.app.DownloadManager;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;

import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.BuildConfig;
import com.pr0gramm.app.Dagger;
import com.pr0gramm.app.R;
import com.pr0gramm.app.services.Update;
import com.pr0gramm.app.services.UpdateChecker;
import com.pr0gramm.app.ui.DialogBuilder;
import com.pr0gramm.app.ui.base.BaseAppCompatActivity;
import com.pr0gramm.app.ui.base.BaseDialogFragment;
import com.trello.rxlifecycle.android.ActivityEvent;

import org.joda.time.Instant;

import javax.inject.Inject;

import rx.Observable;
import rx.functions.Action0;
import rx.functions.Actions;

import static com.pr0gramm.app.ui.fragments.BusyDialogFragment.busyDialog;
import static org.joda.time.Duration.standardHours;
import static org.joda.time.Instant.now;

/**
 */
public class UpdateDialogFragment extends BaseDialogFragment {
    @Inject
    DownloadManager downloadManager;

    @Inject
    SharedPreferences sharedPreferences;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Update update = getArguments().getParcelable("update");
        return update != null
                ? updateAvailableDialog(update)
                : noNewUpdateDialog();
    }

    private Dialog updateAvailableDialog(final Update update) {
        return DialogBuilder.start(getActivity())
                .content(getString(R.string.new_update_available, update.changelog()))
                .positive(R.string.download, () -> UpdateChecker.download(getActivity(), update))
                .negative(R.string.ignore)
                .build();
    }

    private Dialog noNewUpdateDialog() {
        return DialogBuilder.start(getActivity())
                .content(R.string.no_new_update)
                .positive()
                .build();
    }

    public static UpdateDialogFragment newInstance(Update update) {
        Bundle bundle = new Bundle();
        bundle.putParcelable("update", update);

        UpdateDialogFragment dialog = new UpdateDialogFragment();
        dialog.setArguments(bundle);
        return dialog;
    }

    /**
     * Check for updates and shows a new {@link UpdateDialogFragment}
     * if an update could be found.
     *
     * @param activity The activity that starts this update check.
     */
    public static void checkForUpdates(BaseAppCompatActivity activity, boolean interactive) {
        SharedPreferences shared = Dagger.appComponent(activity).sharedPreferences();

        if (!interactive && !BuildConfig.DEBUG) {
            Instant last = new Instant(shared.getLong(KEY_LAST_UPDATE_CHECK, 0));
            if (last.isAfter(now().minus(standardHours(1))))
                return;
        }

        // Action to store the last check time
        Action0 storeCheckTime = () -> shared.edit()
                .putLong(KEY_LAST_UPDATE_CHECK, now().getMillis())
                .apply();

        // show a busy-dialog or not?
        Observable.Operator<Update, Update> busyOperator =
                interactive ? busyDialog(activity) : NOOP;

        // do the check
        new UpdateChecker(activity).check()
                .onErrorResumeNext(Observable.empty())
                .defaultIfEmpty(null)
                .compose(activity.bindUntilEventAsync(ActivityEvent.DESTROY))
                .lift(busyOperator)
                .doAfterTerminate(storeCheckTime)
                .subscribe(update -> {
                    if (interactive || update != null) {
                        UpdateDialogFragment dialog = newInstance(update);
                        dialog.show(activity.getSupportFragmentManager(), null);
                    }
                }, Actions.empty());
    }

    private static final String KEY_LAST_UPDATE_CHECK = "UpdateDialogFragment.lastUpdateCheck";
    private static final Observable.Operator<Update, Update> NOOP = subscriber -> subscriber;

    @Override
    protected void injectComponent(ActivityComponent activityComponent) {
        activityComponent.inject(this);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/dialogs/UpdateDialogFragment.kt
package com.pr0gramm.app.ui.dialogs

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import androidx.core.content.edit
import androidx.fragment.app.FragmentManager
import com.pr0gramm.app.*
import com.pr0gramm.app.services.SingleShotService
import com.pr0gramm.app.services.Update
import com.pr0gramm.app.services.UpdateChecker
import com.pr0gramm.app.ui.base.BaseAppCompatActivity
import com.pr0gramm.app.ui.base.BaseDialogFragment
import com.pr0gramm.app.ui.base.launchWhenStarted
import com.pr0gramm.app.ui.dialog
import com.pr0gramm.app.util.Linkify
import com.pr0gramm.app.util.arguments
import com.pr0gramm.app.util.di.injector
import com.pr0gramm.app.util.di.instance
import com.pr0gramm.app.util.trace
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.withContext
import kotlin.coroutines.coroutineContext

/**
 */
class UpdateDialogFragment : BaseDialogFragment("UpdateDialogFragment") {
    private val singleShotService: SingleShotService by instance()

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val update = arguments?.getParcelable<Update?>("update")
        return update?.let { updateAvailableDialog(it) } ?: noNewUpdateDialog()
    }

    private fun updateAvailableDialog(update: Update): Dialog {
        var content = getString(R.string.new_update_available, update.changelog)

        return dialog(this) {
            positive(R.string.install_update) {
                singleShotService.markAsDoneOnce("update:${update.version}")
                UpdateChecker.download(requireActivity(), update)
            }

            if (!singleShotService.isFirstTime("update:${update.version}")) {
                neutral(R.string.download_manually) {
                    startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(update.apk)))
                }

                content += "\n\n" + getString(R.string.new_update_download)
            }

            content(Linkify.linkify(requireContext(), content))
        }
    }

    private fun noNewUpdateDialog(): Dialog {
        return dialog(this) {
            content(R.string.no_new_update)
            positive()
        }
    }

    companion object {
        private val logger = Logger("Update")

        private fun newInstance(update: Update?): UpdateDialogFragment {
            return UpdateDialogFragment().arguments {
                putParcelable("update", update)
            }
        }

        suspend fun checkForUpdatesInBackground(context: Context, fm: FragmentManager) {
            val prefs = context.injector.instance<SharedPreferences>()

            // only check once an hour.
            if (!BuildConfig.DEBUG) {
                val last = Instant(prefs.getLong(KEY_LAST_UPDATE_CHECK, 0))
                if (last.isAfter(Instant.now() - Duration.hours(1)))
                    return
            }

            val update = withContext(Dispatchers.Default) {
                UpdateChecker().queryAll()
            }

            // remember that we've checked
            prefs.edit {
                putLong(KEY_LAST_UPDATE_CHECK, Instant.now().millis)
            }

            // ignore errors
            if (update is UpdateChecker.Response.Error) {
                logger.warn(update.err) { "Ignoring error during update" }
                return
            }

            if (coroutineContext[Job]?.isCompleted == true) {
                logger.warn { "coroutineContext.job is completed, but still in coroutine?" }
                return
            }

            if (update is UpdateChecker.Response.UpdateAvailable) {
                if (!fm.isStateSaved) {
                    newInstance(update.update).show(fm, null)
                }
            }
        }

        fun checkForUpdatesInteractive(activity: BaseAppCompatActivity) {
            trace { "checkForUpdates" }

            val prefs = activity.applicationContext.injector.instance<SharedPreferences>()

            activity.launchWhenStarted(busyIndicator = true) {
                val update = UpdateChecker().queryAll()

                // remember that we've checked
                prefs.edit {
                    putLong(KEY_LAST_UPDATE_CHECK, Instant.now().millis)
                }

                // let the activity handle the error
                if (update is UpdateChecker.Response.Error) {
                    throw update.err
                }

                val up = (update as? UpdateChecker.Response.UpdateAvailable)?.update
                newInstance(up).show(activity.supportFragmentManager, null)
            }
        }

        private const val KEY_LAST_UPDATE_CHECK = "UpdateDialogFragment.lastUpdateCheck"
    }
}
